### PR TITLE
add a cluster to mange the nodes and mock a real-world network(#187)

### PR DESCRIPTION
**Pull request by xzhangxian1008**
This PR simply implements a cluster to manage nodes and mock a real-world network.

So far, the Cluster could:
- block nodes to mock the network partition
- discard or delay a rpc when the network is set to unreliable
- help the user to check the leader number

TODO for Cluster:
- tests for Cluster should be independent of the raft which may impact the Cluster's tests
- shutdown nodes which will stop sending rpc and lose all the tmp data
- check leader in the same term
- mocking network is needed for HandleRequestPullLogs

Cluster's implementation details:

At the beginning, it creates n raft nodes and n proxy nodes. Each raft node will establish connection with all proxy nodes instead of connecting to other nodes, sees the proxy node as like a raft node and always send messages to the proxy nodes. Proxy nodes are responsible for handling these messages from raft nodes. It will discard or drop the message according to the current context.

rwlock.h:

Since all the proxy nodes need to read the network config when handling every rpc message, we can't simply use the std::mutex which will significiently degrade the concurrent performance.

mock_state_machine.h:

I suppose that this state machine may easy to use in the tests.

### CODE REVIEW

**Comment by jovany-wang** [comment]
Could you add some description for this PR?

File: CMakeLists.txt [hunk 1]
```diff
@@ -72,3 +72,4 @@ RAFTCPP_DEFINE_TEST(test_log_manager SRC_FILES tests/blocking_queue_test.cc LINK
 RAFTCPP_DEFINE_TEST(test_nodeid SRC_FILES tests/nodeid_test.cc LINKS common_lib)
 RAFTCPP_DEFINE_TEST(test_node_elect SRC_FILES tests/node_elect_test.cc LINKS common_lib node_lib)
 RAFTCPP_DEFINE_TEST(logging_check_test SRC_FILES tests/logging_check_test.cc LINKS common_lib node_lib)
+RAFTCPP_DEFINE_TEST(cluster_test SRC_FILES tests/cluster_test.cc LINKS common_lib node_lib)
```
Comment (jovany-wang) [inline CMakeLists.txt:75]: Add a new line?

File: src/node/node.h [hunk 2]
```diff
@@ -64,7 +64,10 @@ class RaftNode : public rpc::NodeService {
 
     void OnHeartbeat(const boost::system::error_code &ec, string_view data);
 
-    RaftState GetCurrState() const { return curr_state_; }
+    RaftState GetCurrState() {
+        std::lock_guard<std::recursive_mutex> guard{mutex_};
```
Comment (jovany-wang) [inline src/node/node.h:68]: I prefer not recursive_mutex here.

File: tests/cluster_test.cc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    // unblock node
+    cl.UnblockNode(second_leader[0]);
+    blocked_nodes = cl.GetBlockedNodes();
+    REQUIRE_EQ(blocked_nodes.size(), 0);
+
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+
+    // ensure it's a stable cluster
+    REQUIRE_EQ(cl.CheckOneLeader(), true);
+}
```
Comment (jovany-wang) [inline tests/cluster_test.cc:70]: Add a new line.

File: tests/rwlock.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <climits>
+#include <condition_variable>
+#include <mutex>
+
+class ReaderWriterLock {
```
Comment (jovany-wang) [inline tests/rwlock.h:7]: Hm, can you explain a bit why we need to implement a read write lock?

Comment (xzhangxian1008) [reply tests/rwlock.h:7]: In order to decide how to treat a rpc, discard or delay, proxy nodes have to get the current network config.

How they get the config? 

They share the NetworkConfig with a shared_ptr<NetworkConfig> and read the NetworkConfig to get the current network config. At the same time, the user may modify the network config and we need to protect the NetworkConfig with a lock.

why rwlock?

Suppose we have 5 proxy nodes and there are 5 rpc messages reach to these proxy nodes at the same time. Only one proxy node could forward the rpc when we simply use a std::mutex. But I hope these 5 proxy nodes could forward the rpc at the same time.

File: CMakeLists.txt (hunk 1 above)
Comment (xzhangxian1008) [reply CMakeLists.txt:75]: just add a test

File: tests/rwlock.h (hunk 4 above)
Comment (jovany-wang) [reply tests/rwlock.h:7]: We have supported C++17 https://github.com/purecpp-org/raftcpp/blob/master/CMakeLists.txt#L3

RWlock is supported in std. It's better to use it in C++ std way.

File: src/node/node.h (hunk 2 above)
Comment (qicosmos) [reply src/node/node.h:68]: I think std::mutex is ok now.

File: tests/rwlock.h (hunk 4 above)
Comment (qicosmos) [reply tests/rwlock.h:7]: Yes, std::shared_mutex from c++17.
https://en.cppreference.com/w/cpp/thread/shared_mutex

Comment (xzhangxian1008) [reply tests/rwlock.h:7]: > We have supported C++17 https://github.com/purecpp-org/raftcpp/blob/master/CMakeLists.txt#L3
> 
> RWlock is supported in std. It's better to use it in C++ std way.

I agree it!

File: src/node/node.h (hunk 2 above)
Comment (xzhangxian1008) [reply src/node/node.h:68]: > I prefer not recursive_mutex here.

+1

File: tests/util.h [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+
+#include "common/config.h"
+#include "common/randomer.h"
+#include "log_manager/log_entry.h"
+#include "mock_state_machine.h"
+#include "node/node.h"
+#include "rest_rpc/rpc_server.h"
+#include "rpc/services.h"
+
+const int DEFAULT_MAX_DELAY = 3000;  // ms
```
Comment (jovany-wang) [inline tests/util.h:23]:
```suggestion
const int DEFAULT_MAX_DELAY_MS = 3000;  // ms
```

File: tests/util.h [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+
+using RpcServ = rest_rpc::rpc_service::rpc_server;
+using RpcConn = rest_rpc::rpc_service::rpc_conn;
```
Comment (jovany-wang) [inline tests/util.h:26]: Great

File: tests/util.h [hunk 7]
```diff
@@ -0,0 +1,613 @@
+#pragma once
+
```
Comment (jovany-wang) [inline tests/util.h:2]: Add a util.cc file ?

**Review by jovany-wang — APPROVED** [review]
LGTM

**Comment by jovany-wang** [comment]
@xzhangxian1008  Can you address the rest comments and then we're going to merge this PR.

**Comment by xzhangxian1008** [comment]
> @xzhangxian1008 Can you address the rest comments and then we're going to merge this PR.

done

**Pull request merged**